pkg/apis/networking/validation/validation_test.go

```diff
@@ -1853,3 +1853,209 @@ func TestValidateIngressStatusUpdate(t *testing.T) {
 		}
 	}
 }
+
+func makeValidClusterCIDRConfig() *networking.ClusterCIDRConfig {
+	return &networking.ClusterCIDRConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            "foo",
+			ResourceVersion: "9",
+		},
+		Spec: networking.ClusterCIDRConfigSpec{
+			IPv4: &networking.CIDRConfig{
+				CIDR:            "10.1.0.0/16",
+				PerNodeMaskSize: int32(24),
+			},
+			IPv6: &networking.CIDRConfig{
+				CIDR:            "fd00:1:1::/64",
+				PerNodeMaskSize: int32(120),
+			},
+			NodeSelector: &api.NodeSelector{
+				NodeSelectorTerms: []api.NodeSelectorTerm{
+					{
+						MatchExpressions: []api.NodeSelectorRequirement{
+							{
+								Key:      "foo",
+								Operator: api.NodeSelectorOpIn,
+								Values:   []string{"bar"},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+type cccTweak func(ccc *networking.ClusterCIDRConfig)
+
+func makeClusterCIDRConfigCustom(tweaks ...cccTweak) *networking.ClusterCIDRConfig {
+	ccc := makeValidClusterCIDRConfig()
+	for _, fn := range tweaks {
+		fn(ccc)
+	}
+	return ccc
+}
+
+func makeCIDRConfig(cidr string, perNodeMaskSize int32) *networking.CIDRConfig {
+	return &networking.CIDRConfig{
+		CIDR:            cidr,
+		PerNodeMaskSize: perNodeMaskSize,
+	}
+}
+
+func makeNodeSelector(key string, op api.NodeSelectorOperator, values []string) *api.NodeSelector {
+	return &api.NodeSelector{
+		NodeSelectorTerms: []api.NodeSelectorTerm{
+			{
+				MatchExpressions: []api.NodeSelectorRequirement{
+					{
+						Key:      key,
+						Operator: op,
+						Values:   values,
+					},
+				},
+			},
+		},
+	}
+}
+
+func TestValidateClusterCIDRConfig(t *testing.T) {
+	// Tweaks used below.
+	setIPv4CIDR := func(cidrConfig *networking.CIDRConfig) cccTweak {
+		return func(ccc *networking.ClusterCIDRConfig) {
+			ccc.Spec.IPv4 = cidrConfig
+		}
+	}
+
+	setIPv6CIDR := func(cidrConfig *networking.CIDRConfig) cccTweak {
+		return func(ccc *networking.ClusterCIDRConfig) {
+			ccc.Spec.IPv6 = cidrConfig
+		}
+	}
+
+	setNodeSelector := func(nodeSelector *api.NodeSelector) cccTweak {
+		return func(ccc *networking.ClusterCIDRConfig) {
+			ccc.Spec.NodeSelector = nodeSelector
+		}
+	}
+
+	validNodeSelector := makeNodeSelector("foo", api.NodeSelectorOpIn, []string{"bar"})
+
+	successCases := map[string]*networking.ClusterCIDRConfig{
+		"valid IPv6 only ClusterCIDRConfig":                      makeClusterCIDRConfigCustom(setIPv4CIDR(nil)),
+		"valid IPv4 only ClusterCIDRConfig":                      makeClusterCIDRConfigCustom(setIPv6CIDR(nil)),
+		"valid DualStack ClusterCIDRConfig with no NodeSelector": makeClusterCIDRConfigCustom(setNodeSelector(nil)),
+		"32 - IPv4.PerNodeMaskSize == 128 - IPv6.PerNodeMaskSize": makeClusterCIDRConfigCustom(
```

Comment:
the outcome of `makeValidClusterCIDRConfig(...)` is already valid without any `custom`. also calling `makeClusterCIDRConfigCustom(...)` with no tweaks should yield into valid `ccc` object. it will also be great to have it as the first test case.

Reply:
Done